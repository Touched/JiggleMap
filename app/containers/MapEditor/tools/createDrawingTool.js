/* @flow */

import * as React from 'react';
import { combineReducers } from 'redux';
import invariant from 'invariant';

import type { Dispatch, Layer, Message, Meta, Object, ReactMouseEvent, Tool } from './types';

export const DRAW_START = 'jigglemap/MapEditor/DrawingTool/DRAW_START';
export const DRAW = 'jigglemap/MapEditor/DrawingTool/DRAW';
export const DRAW_END = 'jigglemap/MapEditor/DrawingTool/DRAW_END';

export const initialState = {
  startingPosition: null,
  currentPosition: null,
  clientOffset: null,
  currentBlock: 0,
};

type Position = {
  x: number;
  y: number;
};

type DrawingToolState = {
  startingPosition: ?Position;
  currentPosition: ?Position;
  clientOffset: ?Position;
  currentBlock: number;
};

type DrawStartAction = {
  type: typeof DRAW_START;
  clientOffset: Position;
  position: Position;
  object: Object;
};

type DrawAction = {
  type: typeof DRAW;
  position: Position;
};

type DrawEndAction = {
  type: typeof DRAW_END;
};

type Action = DrawStartAction | DrawAction | DrawEndAction;

type CombinedState<State> = {
  drawing: DrawingToolState;
  tool: State;
};

type DrawingTool<State> = {
  id: string;
  name: Message;
  description: Message;
  layers: Array<Layer>;
  icon: React.Element<*>;
  component: React.ComponentType<any>;
  reducer?: (state: CombinedState<State>, action: Action) => CombinedState<State>;
  getCursorForObject: (object: Object) => string;
  onDrawStart: (object: Object, position: Position, state: State, tabDispatch: Dispatch) => void;
  onDraw: (position: Position, state: State, tabDispatch: Dispatch) => void;
  onDrawEnd: (state: State, tabDispatch: Dispatch) => void;
};

function toGridCoordinates(x, y) {
  return {
    x: Math.floor(x / 16),
    y: Math.floor(y / 16),
  };
}

function reducer(state: DrawingToolState = initialState, action: Action): DrawingToolState {
  switch (action.type) {
    case DRAW_START:
      return {
        ...state,
        clientOffset: action.clientOffset,
        startingPosition: action.position,
        currentPosition: action.position,
        object: action.object,
      };
    case DRAW:
      return {
        ...state,
        currentPosition: action.position,
      };
    case DRAW_END:
      return {
        ...state,
        startingPosition: null,
      };
    default:
      (action: empty); // eslint-disable-line no-unused-expressions
      return state;
  }
}

export default function createDrawingTool<T: Object>(definition: DrawingTool<T>): Tool<CombinedState<T>> {
  type State = CombinedState<T>;

  const combinedReducer = combineReducers({
    drawing: reducer,
    tool: definition.reducer || ((state = {}) => state),
  });

  return {
    ...definition,
    getCursorForObject: (object) => object.type === 'main-map' ? definition.getCursorForObject(object) : 'auto',
    onMouseDown(object: Object, state: State, meta: Meta, tabDispatch: Dispatch, mouseEvent: ReactMouseEvent) {
      if (object.type === 'main-map') {
        const { nativeEvent: { offsetX, offsetY }, clientX, clientY } = mouseEvent;
        const position = toGridCoordinates(offsetX, offsetY);

        tabDispatch({
          type: DRAW_START,
          clientOffset: {
            x: (clientX / meta.zoomLevel) - offsetX,
            y: (clientY / meta.zoomLevel) - offsetY,
          },
          position,
          object,
        });

        const updatedState = {
          ...state,
          drawing: {
            ...state.drawing,
            startingPosition: position,
          },
        };

        this.onDrawStart(position, updatedState, tabDispatch);
        this.onDraw(position, updatedState, tabDispatch);
      }
    },
    onMouseMove(state: State, meta: Meta, tabDispatch: Dispatch, mouseEvent: ReactMouseEvent) {
      if (state.drawing.startingPosition) {
        const { clientOffset, currentPosition } = state.drawing;

        invariant(clientOffset, 'mousedown event did not set the clientOffset');
        invariant(currentPosition, 'mousedown event did not set the currentPosition');

        const x = (mouseEvent.clientX / meta.zoomLevel) - clientOffset.x;
        const y = (mouseEvent.clientY / meta.zoomLevel) - clientOffset.y;
        const position = toGridCoordinates(x, y);

        if (currentPosition.x !== position.x || currentPosition.y !== position.y) {
          tabDispatch({
            type: DRAW,
            position,
          });

          this.onDraw(position, state, tabDispatch);
        }
      }
    },
    onMouseUp(state: State, meta: Meta, tabDispatch: Dispatch) {
      if (state.drawing.startingPosition) {
        tabDispatch({
          type: DRAW_END,
        });

        this.onDrawEnd(state, tabDispatch);
      }
    },
    reducer: combinedReducer,
  };
}
