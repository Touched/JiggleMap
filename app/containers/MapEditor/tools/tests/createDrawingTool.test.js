import createDrawingTool, { DRAW_START, DRAW, DRAW_END, initialState } from '../createDrawingTool';

describe('createDrawingTool', () => {
  let drawingTool;
  const reducer = jest.fn((state = { foo: 'bar' }) => state);

  beforeEach(() => {
    drawingTool = createDrawingTool({
      onDrawStart: jest.fn(),
      onDraw: jest.fn(),
      onDrawEnd: jest.fn(),
      getCursorForObject: () => 'pointer',
      reducer,
    });
  });

  const object = {
    type: 'main-map',
  };

  const meta = {
    zoomLevel: 2,
  };

  const dispatch = jest.fn();

  it('returns a cursor only over the main map', () => {
    expect(drawingTool.getCursorForObject({ type: 'main-map' })).toEqual('pointer');
    expect(drawingTool.getCursorForObject({ type: 'other' })).toEqual('auto');
  });

  describe('reducer', () => {
    it('sets the intial state', () => {
      expect(drawingTool.reducer(undefined, {})).toMatchSnapshot();
      expect(createDrawingTool({}).reducer(undefined, {})).toMatchSnapshot();
    });

    describe('DRAW_START', () => {
      it('stores the information', () => {
        const position = {
          x: 10,
          y: 5,
        };

        const clientOffset = { x: 156, y: 167 };

        const state = drawingTool.reducer({}, {
          type: DRAW_START,
          object,
          position,
          clientOffset,
        });

        expect(state).toMatchObject({
          drawing: {
            clientOffset,
            startingPosition: position,
            currentPosition: position,
            object,
          },
        });
      });
    });

    describe('DRAW', () => {
      it('updates the currentPosition', () => {
        const position = { x: 10, y: 5 };
        const state = drawingTool.reducer({}, {
          type: DRAW,
          position,
        });

        expect(state).toMatchObject({
          drawing: {
            currentPosition: position,
          },
        });
      });
    });

    describe('DRAW_END', () => {
      it('sets the startingPosition to null', () => {
        const state = drawingTool.reducer({}, {
          type: DRAW_END,
        });

        expect(state).toMatchObject({
          drawing: {
            startingPosition: null,
          },
        });
      });
    });

    describe('default', () => {
      it('passes the action through to the child reducer', () => {
        const action = {
          type: 'foo',
        };

        const state = drawingTool.reducer(undefined, action);

        expect(reducer).toHaveBeenCalledWith(undefined, action);
        expect(state).toEqual({
          drawing: initialState,
          tool: {
            foo: 'bar',
          },
        });
      });
    });
  });

  describe('draw start', () => {
    const event = {
      nativeEvent: {
        offsetX: 160,
        offsetY: 80,
      },
      clientX: 456,
      clientY: 186,
    };

    const position = {
      x: 10,
      y: 5,
    };

    const state = {
      drawing: {
        foo: 'bar',
      },
    };

    const updatedState = {
      ...state,
      drawing: {
        ...state.drawing,
        startingPosition: position,
      },
    };

    it('it does not call onDrawStart when onMouseDown is called with a non main-map object', () => {
      drawingTool.onMouseDown({});
      expect(drawingTool.onDrawStart).not.toHaveBeenCalled();
    });

    it('calls onDrawStart when onMouseDown is called with a main-map object', () => {
      drawingTool.onMouseDown(object, state, meta, dispatch, event);
      expect(drawingTool.onDrawStart).toHaveBeenCalledWith(position, updatedState, dispatch);
    });

    it('calls onDraw when onMouseDown is called with a main-map object', () => {
      drawingTool.onMouseDown(object, state, meta, dispatch, event);
      expect(drawingTool.onDraw).toHaveBeenCalledWith(position, updatedState, dispatch);
    });

    it('dispatches a DRAW_START action', () => {
      drawingTool.onMouseDown(object, updatedState, meta, dispatch, event);
      expect(dispatch).toHaveBeenLastCalledWith({
        type: DRAW_START,
        clientOffset: {
          x: 68,
          y: 13,
        },
        position,
        object,
      });
    });
  });

  describe('draw', () => {
    const event = {
      nativeEvent: {
        offsetX: 160,
        offsetY: 80,
      },
      clientX: 456,
      clientY: 186,
    };

    const position = {
      x: 10,
      y: 5,
    };

    const state = {
      drawing: {
        clientOffset: {
          x: 115,
          y: 76,
        },
        startingPosition: position,
        currentPosition: position,
      },
    };

    it('does not call onDraw if there was no onDrawStart', () => {
      drawingTool.onMouseMove({ drawing: {} }, meta, dispatch, event);
      expect(drawingTool.onDraw).not.toHaveBeenCalled();
    });

    it('calls onDraw if there was a onDrawStart', () => {
      drawingTool.onMouseMove(state, meta, dispatch, event);
      expect(drawingTool.onDraw).toHaveBeenCalledWith({
        x: 7,
        y: 1,
      }, state, dispatch);
    });

    it('does not call onDraw if the cursor has not moved to a different grid cell', () => {
      const newState = {
        ...state,
        drawing: {
          ...state.drawing,
          startingPosition: position,
          currentPosition: { x: 7, y: 1 },
        },
      };

      drawingTool.onMouseMove(newState, meta, dispatch, event);
      expect(drawingTool.onDraw).not.toHaveBeenCalled();
    });

    it('dispatches a DRAW action', () => {
      drawingTool.onMouseMove(state, meta, dispatch, event);
      expect(dispatch).toHaveBeenLastCalledWith({
        type: DRAW,
        position: { x: 7, y: 1 },
      });
    });
  });

  describe('draw end', () => {
    const position = {};

    const state = {
      drawing: {
        startingPosition: position,
      },
    };

    it('does not call onDrawEnd if there was no onDrawStart', () => {
      drawingTool.onMouseUp({ drawing: {} }, meta, dispatch);
      expect(drawingTool.onDrawEnd).not.toHaveBeenCalled();
    });

    it('calls onDrawEnd if there was a onDrawStart', () => {
      drawingTool.onMouseUp(state, meta, dispatch);
      expect(drawingTool.onDrawEnd).toHaveBeenCalledWith(state, dispatch);
    });

    it('dispatches a DRAW_END action', () => {
      drawingTool.onMouseUp(state, meta, dispatch);
      expect(dispatch).toHaveBeenLastCalledWith({
        type: DRAW_END,
      });
    });
  });
});
