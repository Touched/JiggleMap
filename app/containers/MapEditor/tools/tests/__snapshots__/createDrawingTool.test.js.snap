// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDrawingTool reducer sets the intial state 1`] = `
Object {
  "drawing": Object {
    "clientOffset": null,
    "currentBlock": 0,
    "currentPosition": null,
    "startingPosition": null,
  },
  "tool": Object {
    "foo": "bar",
  },
}
`;

exports[`createDrawingTool reducer sets the intial state 2`] = `
Object {
  "drawing": Object {
    "clientOffset": null,
    "currentBlock": 0,
    "currentPosition": null,
    "startingPosition": null,
  },
  "tool": Object {},
}
`;
